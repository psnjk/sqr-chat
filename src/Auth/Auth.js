import React, { useState } from "react";
import {
  signInWithEmailAndPassword,
  createUserWithEmailAndPassword,
  sendPasswordResetEmail,
  GoogleAuthProvider,
  signInWithRedirect,
} from "firebase/auth";
import { collection, addDoc } from "firebase/firestore";

import cn from "./Auth.module.scss";

const Auth = ({ auth, db }) => {
  const provider = new GoogleAuthProvider();

  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [name, setName] = useState("");

  const [status, setStatus] = useState("");

  if (status === "reset") {
    return (
      <>
        <div className={cn.root} data-testid="reset">
          <input
            type="text"
            className={cn.input}
            value={email}
            onChange={(e) => setEmail(e.target.value)}
            placeholder="E-mail Address"
          />
          <button
            className={cn.button}
            onClick={() => sendPasswordReset(auth, email)}
          >
            Send password reset email
          </button>
        </div>
        <button
          className={cn.buttonReg}
          onClick={() => setStatus("login")}
          data-testid="toLogin-id"
        >
          Log In
        </button>
      </>
    );
  }
  if (status === "reg") {
    return (
      <div className={cn.root} data-testid="reg">
        <input
          type="text"
          className={cn.input}
          value={name}
          onChange={(e) => setName(e.target.value)}
          placeholder="Full Name"
        />
        <input
          type="text"
          className={cn.input}
          value={email}
          onChange={(e) => setEmail(e.target.value)}
          placeholder="E-mail Address"
        />
        <input
          type="password"
          className={cn.input}
          value={password}
          onChange={(e) => setPassword(e.target.value)}
          placeholder="Password"
        />
        <button
          className={cn.button}
          onClick={() => register(auth, db, name, email, password)}
        >
          Register
        </button>
        <button
          className={cn.buttonReg}
          onClick={() => setStatus("login")}
          data-testid="toLogin-id"
        >
          Already have an account?
        </button>
      </div>
    );
  }
  return (
    <>
      <div className={cn.root} data-testid="login">
        <input
          type="text"
          className={cn.input}
          value={email}
          onChange={(e) => setEmail(e.target.value)}
          placeholder="E-mail Address"
        />
        <input
          type="password"
          className={cn.input}
          value={password}
          onChange={(e) => setPassword(e.target.value)}
          placeholder="Password"
        />
        <button
          className={cn.button}
          onClick={() => logInWithEmailAndPassword(auth, email, password)}
        >
          Login
        </button>
        <button
          className={cn.buttonReg}
          onClick={() => setStatus("reg")}
          data-testid="toReg-id"
        >
          Do not have an account?
        </button>
        <button
          className={cn.buttonReg}
          onClick={() => setStatus("reset")}
          data-testid="toReset-id"
        >
          Forgot password
        </button>
        <button onClick={() => signInGoogle(auth, provider)}>Google</button>
      </div>
    </>
  );
};

export default Auth;

export const logInWithEmailAndPassword = async (auth, email, password) => {
  try {
    await signInWithEmailAndPassword(auth, email, password);
  } catch (err) {
    // console.error(err);
    // alert(err.message);
    return;
  }
};

export const registerWithEmailAndPassword = async (
  auth,
  db,
  name,
  email,
  password
) => {
  try {
    const res = await createUserWithEmailAndPassword(auth, email, password);
    const user = res.user;
    await addDoc(collection(db, "users"), {
      uid: user.uid,
      name,
      authProvider: "local",
      email,
    });
  } catch (err) {
    // console.error(err);
    // alert(err.message);
    return;
  }
};
export const register = (auth, db, name, email, password) => {
  if (!name) {
    return;
  }
  registerWithEmailAndPassword(auth, db, name, email, password);
};

export const sendPasswordReset = async (auth, email) => {
  try {
    await sendPasswordResetEmail(auth, email);
    alert("Password reset link sent!");
  } catch (err) {
    // console.error(err);
    // alert(err.message);
    return;
  }
};

export const signInGoogle = (auth, provider) => {
  if (!auth || !provider) return;
  signInWithRedirect(auth, provider);
};
