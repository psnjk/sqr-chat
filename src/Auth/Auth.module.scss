.root{
	display: flex;
	flex-direction: column;
	align-items: center;
}
.input{
	border: 0px;
	border-radius: 5px;
	margin: 10px;
	padding: 5px 10px;
	width: 250px;
	height: 20px;
}
.button{
	border: 0px;
	border-radius: 5px;
	margin: 10px;
	padding: 5px;
	width: 200px;
}
.buttonReg{
	composes: button;
	color: rgb(87, 87, 87);
}