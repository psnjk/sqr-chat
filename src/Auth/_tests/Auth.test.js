import React from "react";
import { act } from "react-dom/test-utils";
import renderer from "react-test-renderer";
import { render, cleanup, screen, waitFor } from "@testing-library/react";
import { wrapper, shallow } from "enzyme";
import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import * as firebaseFunctions from "firebase/auth";
import Auth, * as AuthFunc from "../Auth";
import userEvent from "@testing-library/user-event";

Enzyme.configure({ adapter: new Adapter() });

afterEach(cleanup);

it("it renders", () => {
  render(<Auth />);
});

it("it renders w props", () => {
  render(<Auth auth={{}} db={{}} />);
});

// it("it renders error", () => {
//   const { getByTestId } = render(<Auth />);
//   expect(getByTestId("error")).not.toBeNull();
// });

it("it renders login", () => {
  const { getByTestId } = render(<Auth />);
  expect(getByTestId("login")).not.toBeNull();
  expect(screen.queryByTestId("reset")).not.toBeInTheDocument();
  expect(screen.getByTestId("toReg-id")).toBeInTheDocument();
  // expect(screen.getByTestId("reg")).not.toBeInTheDocument();
});

it("it renders not reset", () => {
  const { getByTestId } = render(<Auth />);
  expect(screen.queryByTestId("reset")).not.toBeInTheDocument();
});

it("it renders not reg", () => {
  const { getByTestId } = render(<Auth />);
  expect(getByTestId("login")).not.toBeNull();
  expect(screen.queryByTestId("reg")).not.toBeInTheDocument();
  // expect(screen.getByTestId("reg")).not.toBeInTheDocument();
});

it("render reg", async () => {
  render(<Auth />);
  userEvent.click(screen.getByTestId("toReg-id"));
  await waitFor(() => expect(screen.getByTestId("reg")).toBeInTheDocument());
});

it("render reset", async () => {
  render(<Auth />);
  userEvent.click(screen.getByTestId("toReset-id"));
  await waitFor(() => expect(screen.getByTestId("reset")).toBeInTheDocument());
});

it("back to login from reset", async () => {
  render(<Auth />);
  userEvent.click(screen.getByTestId("toReset-id"));
  await waitFor(() => expect(screen.getByTestId("reset")).toBeInTheDocument());
  userEvent.click(screen.getByTestId("toLogin-id"));
  await waitFor(() => expect(screen.getByTestId("login")).toBeInTheDocument());
});

it("back to login from reg", async () => {
  render(<Auth />);
  userEvent.click(screen.getByTestId("toReg-id"));
  await waitFor(() => expect(screen.getByTestId("reg")).toBeInTheDocument());
  userEvent.click(screen.getByTestId("toLogin-id"));
  await waitFor(() =>
    expect(screen.queryByTestId("login")).toBeInTheDocument()
  );
});

it("matches snapshot", () => {
  const tree = renderer.create(<Auth />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("login", () => {
  render(<Auth />);
  userEvent.click(screen.getByTestId("toReg-id"));
});

// test("function testing", () => {
//   const wrapper = shallow(<Auth />);
//   wrapper.instance().logInWithEmailAndPassword();
//   // expect(t).toThrowError();
// });

test("funct call reg", async () => {
  const r = await AuthFunc.register();
  expect(r).toBe(undefined);
});

test("funct reset", async () => {
  const r = await AuthFunc.sendPasswordReset();
  expect(r).toBe(undefined);
});

test("funct reg", async () => {
  const r = await AuthFunc.registerWithEmailAndPassword();
  expect(r).toBe(undefined);
});

test("funct login", async () => {
  const r = await AuthFunc.logInWithEmailAndPassword();
  expect(r).toBe(undefined);
});
test("funct google", async () => {
  const r = await AuthFunc.signInGoogle();
  expect(r).toBe(undefined);
});

// test("register full", async () => {
//   const spy = jest
//     .spyOn(firebaseFunctions, "createUserWithEmailAndPassword")
//     .mockImplementationOnce(() => undefined);
//   const r = await AuthFunc.registerWithEmailAndPassword();
//   expect(r).toBe(undefined);
//   // const { getByTestId } = render(<App />);
//   // expect(getByTestId("error")).toBeInTheDocument();

//   spy.mockRestore();
// });
