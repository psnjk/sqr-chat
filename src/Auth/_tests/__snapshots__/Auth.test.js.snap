// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="root"
  data-testid="login"
>
  <input
    className="input"
    onChange={[Function]}
    placeholder="E-mail Address"
    type="text"
    value=""
  />
  <input
    className="input"
    onChange={[Function]}
    placeholder="Password"
    type="password"
    value=""
  />
  <button
    className="button"
    onClick={[Function]}
  >
    Login
  </button>
  <button
    className="buttonReg"
    data-testid="toReg-id"
    onClick={[Function]}
  >
    Do not have an account?
  </button>
  <button
    className="buttonReg"
    data-testid="toReset-id"
    onClick={[Function]}
  >
    Forgot password
  </button>
  <button
    onClick={[Function]}
  >
    Google
  </button>
</div>
`;
